File: app/app/src/main/java/com/example/hakonsreader/api/requestmodels/ModRequestModel.java
package com.example.hakonsreader.api.requestmodels;

import com.example.hakonsreader.api.RedditApi;
import com.example.hakonsreader.api.enums.Thing;
import com.example.hakonsreader.api.interfaces.OnFailure;
import com.example.hakonsreader.api.interfaces.OnResponse;
import com.example.hakonsreader.api.model.AccessToken;
import com.example.hakonsreader.api.model.RedditComment;
import com.example.hakonsreader.api.model.RedditPost;
import com.example.hakonsreader.api.responses.GenericError;
import com.example.hakonsreader.api.responses.JsonResponseKt;
import com.example.hakonsreader.api.service.ModService;
import com.example.hakonsreader.api.utils.Util;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class ModRequestModel {

    private final AccessToken accessToken;
    private final ModService api;

    public ModRequestModel(AccessToken accessToken, ModService api) {
        this.accessToken = accessToken;
        this.api = api;
    }

    /**
     * Distinguish a comment as mod, and optionally sticky it
     *
     * @param id The ID of the comment to distinguish
     * @param distinguish True to distinguish as mod, false to remove previous distinguish
     * @param sticky True to also sticky the comment. This is only possible on top-level comments
     * @param onResponse Callback for successful requests. Holds the updated comment info
     * @param onFailure Callback for failed requests.
     */
    public void distinguishAsModComment(String id, boolean distinguish, boolean sticky, OnResponse<RedditComment> onResponse, OnFailure onFailure) {
        api.distinguishAsModComment(
                Util.createFullName(Thing.COMMENT, id),
                distinguish ? "yes" : "no",
                sticky,
                RedditApi.API_TYPE
        ).enqueue(new Callback<JsonResponseKt<RedditComment>>() {
            @Override
            public void onResponse(Call<JsonResponseKt<RedditComment>> call, Response<JsonResponseKt<RedditComment>> response) {
                if (response.isSuccessful()) {
                    JsonResponseKt<RedditComment> body = response.body();

                    if (body != null) {
                        RedditComment comment = (RedditComment) body.getResponse().getListings().get(0);
                        onResponse.onResponse(comment);
                    } else {
                        Util.handleHttpErrors(response, onFailure);
                    }
                } else {
                    Util.handleHttpErrors(response, onFailure);
                }
            }

            @Override
            public void onFailure(Call<JsonResponseKt<RedditComment>> call, Throwable t) {
                onFailure.onFailure(new GenericError(-1), t);
            }
        });
    }

    /**
     * Distinguish a post as mod
     *
     * @param id The ID of the comment to distinguish
     * @param distinguish True to distinguish as mod, false to remove previous distinguish
     * @param onResponse Callback for successful requests. Holds the updated comment info
     * @param onFailure Callback for failed requests.
     */
    public void distinguishAsModPost(String id, boolean distinguish, OnResponse<RedditPost> onResponse, OnFailure onFailure) {
        // TODO this
        /*
        api.distinguishAsModPost(
                Util.createFullName(Thing.POST, id),
                distinguish ? "yes" : "no",
                RedditApi.API_TYPE
        ).enqueue(new Callback<JsonResponseKt<RedditPost>>() {
            @Override
            public void onResponse(Call<JsonResponseKt<RedditPost>> call, Response<JsonResponseKt<RedditPost>> response) {
                if (response.isSuccessful()) {
                    JsonResponseKt<RedditPost> body = response.body();

                    if (body != null) {
                        RedditPost post = (RedditPost) body.getResponse().getListings().get(0);
                        onResponse.onResponse(post);
                    } else {
                        Util.handleHttpErrors(response, onFailure);
                    }
                } else {
                    Util.handleHttpErrors(response, onFailure);
                }
            }

            @Override
            public void onFailure(Call<JsonResponseKt<RedditPost>> call, Throwable t) {
                onFailure.onFailure(new GenericError(-1), t);
            }
        });
         */
    }

    /**
     * Sticky or unsticky a post
     *
     * @param id The id of the post
     * @param sticky True to sticky, false to unsticky. If the post is already stickied and this is true,
     *               a 409 Conflict error will occur
     * @param onResponse The callback for successful requests. Nothing is returned here, but the callback
     *                   will be called when the request is successful
     * @param onFailure The callback for failed requests
     */
    public void stickyPost(String id, boolean sticky, OnResponse<Void> onResponse, OnFailure onFailure) {
        // TODO this
        /*
        api.stickyPost(
                Util.createFullName(Thing.POST, id),
                sticky,
                RedditApi.API_TYPE
        ).enqueue(new Callback<JsonResponseKt<RedditPost>>() {
            @Override
            public void onResponse(Call<JsonResponseKt<RedditPost>> call, Response<JsonResponseKt<RedditPost>> response) {
                if (response.isSuccessful()) {
                    JsonResponseKt<RedditPost> body = response.body();

                    if (body != null) {
                        // There is no actual data in the response
                        onResponse.onResponse(null);
                    } else {
                        Util.handleHttpErrors(response, onFailure);
                    }
                } else {
                    Util.handleHttpErrors(response, onFailure);
                }
            }

            @Override
            public void onFailure(Call<JsonResponseKt<RedditPost>> call, Throwable t) {
                onFailure.onFailure(new GenericError(-1), t);
            }
        });
         */
    }
}

File: app/app/src/main/java/com/example/hakonsreader/api/requestmodels/ModRequestModel.kt
package com.example.hakonsreader.api.requestmodels

import com.example.hakonsreader.api.enums.Thing
import com.example.hakonsreader.api.exceptions.InvalidAccessTokenException
import com.example.hakonsreader.api.model.AccessToken
import com.example.hakonsreader.api.model.RedditComment
import com.example.hakonsreader.api.model.RedditPost
import com.example.hakonsreader.api.responses.ApiResponse
import com.example.hakonsreader.api.responses.GenericError
import com.example.hakonsreader.api.service.ModService
import com.example.hakonsreader.api.utils.apiError
import com.example.hakonsreader.api.utils.createFullName
import com.example.hakonsreader.api.utils.verifyLoggedInToken
import java.lang.Exception

/**
 * Request model for performing various mod related actions
 */
class ModRequestModel(
        private val accessToken: AccessToken,
        private val api: ModService
) {


    /**
     * Distinguish a comment as mod, and optionally sticky it
     *
     * @param id The ID of the comment to distinguish
     * @param distinguish True to distinguish as mod, false to remove previous distinguish
     * @param sticky True to also sticky the comment. This is only possible on top-level comments
     */
    suspend fun distinguishAsModComment(id: String, distinguish: Boolean, sticky: Boolean) : ApiResponse<RedditComment> {
        try {
            verifyLoggedInToken(accessToken)
        } catch (e: InvalidAccessTokenException) {
            return ApiResponse.Error(GenericError(-1), InvalidAccessTokenException("Distinguishing/stickying a comment requires a valid access token for a logged in user", e))
        }

        return try {
            val resp = api.distinguishAsModComment(
                    createFullName(Thing.COMMENT, id),
                    if (distinguish) "yes" else "no",
                    sticky,
            )

            val comment = resp.body()?.getListings()?.get(0)
            if (comment != null) {
                ApiResponse.Success(comment)
            } else {
                apiError(resp)
            }
        } catch (e: Exception) {
            ApiResponse.Error(GenericError(-1), e)
        }
    }

    /**
     * Distinguish a post as mod
     *
     * @param id The ID of the post to distinguish
     * @param distinguish True to distinguish as mod, false to remove previous distinguish
     */
    suspend fun distinguishAsModPost(id: String, distinguish: Boolean) : ApiResponse<RedditPost> {
        try {
            verifyLoggedInToken(accessToken)
        } catch (e: InvalidAccessTokenException) {
            return ApiResponse.Error(GenericError(-1), InvalidAccessTokenException("Distinguishing a post requires a valid access token for a logged in user", e))
        }

        return try {
            val resp = api.distinguishAsModPost(
                    createFullName(Thing.POST, id),
                    if (distinguish) "yes" else "no"
            )

            val post = resp.body()?.getListings()?.get(0)
            if (post != null) {
                ApiResponse.Success(post)
            } else {
                apiError(resp)
            }
        } catch (e: Exception) {
            ApiResponse.Error(GenericError(-1), e)
        }
    }


    /**
     * Sticky or unsticky a post
     *
     * @param id The id of the post
     * @param sticky True to sticky, false to unsticky. If the post is already stickied and this is true,
     *               a 409 Conflict error will occur
     * @return No data is returned
     */
    suspend fun stickyPost(id: String, sticky: Boolean) : ApiResponse<Any?> {
        try {
            verifyLoggedInToken(accessToken)
        } catch (e: InvalidAccessTokenException) {
            return ApiResponse.Error(GenericError(-1), InvalidAccessTokenException("Stickying a post requires a valid access token for a logged in user", e))
        }

        return try {
            val resp = api.stickyPost(
                    createFullName(Thing.POST, id),
                    sticky
            )

            if (resp.isSuccessful) {
                ApiResponse.Success(null)
            } else {
                apiError(resp)
            }
        } catch (e: Exception) {
            ApiResponse.Error(GenericError(-1), e)
        }
    }

    /**
     * Ignore reports on a post or comment
     *
     * @see unignoreReports
     */
    suspend fun ignoreReports(thing: Thing, id: String) : ApiResponse<Any?> {
        try {
            verifyLoggedInToken(accessToken)
        } catch (e: InvalidAccessTokenException) {
            return ApiResponse.Error(GenericError(-1), InvalidAccessTokenException("Ignoring reports requires a valid access token for a logged in user", e))
        }

        return try {
            val resp = api.ignoreReports(
                    createFullName(thing, id)
            )

            if (resp.isSuccessful) {
                ApiResponse.Success(null)
            } else {
                apiError(resp)
            }
        } catch (e: Exception) {
            ApiResponse.Error(GenericError(-1), e)
        }
    }

    /**
     * Unignore reports on a post or comment
     *
     * @return No response data is returned
     * @see ignoreReports
     */
    suspend fun unignoreReports(thing: Thing, id: String) : ApiResponse<Any?> {
        try {
            verifyLoggedInToken(accessToken)
        } catch (e: InvalidAccessTokenException) {
            return ApiResponse.Error(GenericError(-1), InvalidAccessTokenException("Unignoring reports requires a valid access token for a logged in user", e))
        }

        return try {
            val resp = api.unignoreReports(
                    createFullName(thing, id)
            )

            if (resp.isSuccessful) {
                ApiResponse.Success(null)
            } else {
                apiError(resp)
            }
        } catch (e: Exception) {
            ApiResponse.Error(GenericError(-1), e)
        }
    }

    /**
     * Lock a post or comment
     *
     * @param id The id of the post/comment
     * @param isPost True if a post is being locked, false if a comment is being locked
     * @return No response data is returned
     * @see unlock
     */
    suspend fun lock(id: String, isPost: Boolean) : ApiResponse<Any?> {
        return lockInternal(if (isPost) Thing.POST else Thing.COMMENT, id, true)
    }

    /**
     * Unlock a post or comment
     *
     * @param id The id of the post/comment
     * @param isPost True if a post is being unlocked, false if a comment is being unlocked
     * @return No response data is returned
     * @see lock
     */
    suspend fun unlock(id: String, isPost: Boolean) : ApiResponse<Any?> {
        return lockInternal(if (isPost) Thing.POST else Thing.COMMENT, id, false)
    }

    private suspend fun lockInternal(thing: Thing, id: String, lock: Boolean) : ApiResponse<Any?> {
        try {
            verifyLoggedInToken(accessToken)
        } catch (e: InvalidAccessTokenException) {
            return ApiResponse.Error(GenericError(-1), InvalidAccessTokenException("Cannot lock/unlock without a valid access token for a logged in user", e))
        }

        val fullname = createFullName(thing, id)

        return try {
            val resp = if (lock) {
                api.lock(fullname)
            } else {
                api.unlock(fullname)
            }

            if (resp.isSuccessful) {
                ApiResponse.Success(null)
            } else {
                apiError(resp)
            }
        } catch (e: Exception) {
            ApiResponse.Error(GenericError(-1), e)
        }
    }
}